src/irr/asset/CGLSLVirtualTexturingBuiltinIncludeLoader.h

```diff
@@ -0,0 +1,224 @@
+{
+public:
+    const char* getVirtualDirectoryName() const override { return "glsl/virtual_texturing/"; }
+
+private:
+	static core::vector<std::string> parseArgumentsFromPath(const std::string& _path)
+	{
+		core::vector<std::string> args;
+
+		std::stringstream ss{ _path };
+		std::string arg;
+		while (std::getline(ss, arg, '/'))
+			args.push_back(std::move(arg));
+
+		return args;
+	}
+    static std::string getVTfunctions(const std::string& _path)
+    {
+		auto args = parseArgumentsFromPath(_path.substr(_path.find_first_of('/')+1, _path.npos));
+		if (args.size()<10u)
+			return {};
+
+		using namespace std::string_literals;
+        std::string s = "#include <irr/builtin/glsl/texture_packer/utils.glsl";
+		for (uint32_t i = 0u; i < 4u; ++i)
+			s += "/" + args[i];
+        s += ">";
+		s += //those are supposed to be used by VT fucntions only (hence extra VT_ pefix)
+			"\n#define irr_glsl_VT_pgtabName "s + args[4] +
+			"\n#define irr_glsl_VT_physPgTexName " + args[5] +
+			"\n#define irr_glsl_VT_getPgTabSzLog2 " + args[6] +
+			"\n#define irr_glsl_VT_getPhysPgTexSzRcp " + args[7] +
+			"\n#define irr_glsl_VT_getVTexSzRcp " + args[8] +
+			"\n#define irr_glsl_VT_layer2pid " + args[9]
+			;
+        s += R"(
+
+#extension GL_EXT_nonuniform_qualifier  : enable
+
+#ifdef IRR_GL_NV_gpu_shader5
+    #define IRR_GL_EXT_nonuniform_qualifier // TODO: we need to overhaul our GLSL preprocessing system to match what SPIRV-Cross actually does
+#endif
+
+#ifndef IRR_GL_EXT_nonuniform_qualifier
+    #extension GL_ARB_gpu_shader_int64      : require
+    #extension GL_ARB_shader_ballot         : require
+#endif
+
+vec3 irr_glsl_vTextureGrad_helper(in uint formatID, in vec3 virtualUV, in int clippedLoD, in int levelInTail)
+{
+    uvec2 pageID = textureLod(irr_glsl_VT_pgtabName,virtualUV,clippedLoD).xy;
+
+	const uint pageTableSizeLog2 = irr_glsl_VT_getPgTabSzLog2(formatID);
+    const float phys_pg_tex_sz_rcp = irr_glsl_VT_getPhysPgTexSzRcp(formatID);
+	// assert that pageTableSizeLog2<23
+
+	// this will work because pageTables are always square and PoT and IEEE754
+	uint thisLevelTableSize = (pageTableSizeLog2-uint(clippedLoD))<<23;
+
+	vec2 tileCoordinate = uintBitsToFloat(floatBitsToUint(virtualUV.xy)+thisLevelTableSize);
+	tileCoordinate = fract(tileCoordinate); // optimize this fract at some point
+	tileCoordinate = uintBitsToFloat(floatBitsToUint(tileCoordinate)+uint((PAGE_SZ_LOG2-levelInTail)<<23));
+    tileCoordinate += packingOffsets[levelInTail];
+	tileCoordinate += vec2(TILE_PADDING,TILE_PADDING);
+	tileCoordinate *= phys_pg_tex_sz_rcp;
+
+	vec3 physicalUV = irr_glsl_unpackPageID(levelInTail!=0 ? pageID.y:pageID.x);
+	physicalUV.xy *= vec2(PAGE_SZ+2*TILE_PADDING)*phys_pg_tex_sz_rcp;
+
+	// add the in-tile coordinate
+	physicalUV.xy += tileCoordinate;
+    return physicalUV;
+}
+
+float irr_glsl_lengthManhattan(vec2 v)
+{
+	v = abs(v);
+    return v.x+v.y;
+}
+float irr_glsl_lengthSq(in vec2 v)
+{
+  return dot(v,v);
+}
```

Comment:
this needs to go somewhere else and needs a definition for each of the `float` to `vec4` types

Reply:
btw overloads are safe now!

Reply:
these functions need to move to a different header

Reply:
still an issue

Reply:
still!